### PR TITLE
[RPC] Headers: minimal HPACK implementation.

### PR DESCRIPTION
**Pull request by kobolog**

### CODE REVIEW

File: include/cocaine/rpc/asio/encoder.hpp [hunk 1]
```diff
@@ -95,29 +95,77 @@ struct encoded_message_t {
...
+        // Message arguments
+
+        type_traits<typename event_traits<Event>::argument_type>::pack(packer,
+            std::forward<Args>(args)...);
+
+        // Optional message metadata
+
+        packer.pack_array(3);
+
+        hpack::msgpack_traits::pack<hpack::headers::trace_id<>>(packer);
```
Comment (antmat) [inline include/cocaine/rpc/asio/encoder.hpp]: How do we pack real values here without header table?

Comment (kobolog) [inline include/cocaine/rpc/asio/encoder.hpp]: encoder.hpack_context?

File: include/cocaine/rpc/asio/encoder.hpp [hunk 2]
```diff
@@ -95,29 +95,77 @@ struct encoded_message_t {
...
+private:
+    // HPACK HTTP/2.0 tables.
+    hpack::header_table_t hpack_context;
 };
 
-struct encoder_t {
-    typedef aux::encoded_message_t message_type;
+template<class Event>
+struct encoded:
+    public aux::unbound_message_t
```
Comment (antmat) [inline include/cocaine/rpc/asio/encoder.hpp]: Why do we need inheritance here?

Comment (kobolog) [inline include/cocaine/rpc/asio/encoder.hpp]: https://github.com/cocaine/cocaine-core/blob/kobolog/pr-refactoring/3Hren-pr/hpack/include/cocaine/rpc/upstream.hpp#L50

**Comment by antmat** [comment]
LGTM, except CNF breakage, but it can be fixed separately.

**Pull request merged**